### PR TITLE
task ordering, remove unused jobState task arrays

### PR DESCRIPTION
**Pull request by JeanetteBruno**
- the load_based_scheduler was not selecting tasks in decreasing duration order.  load_based_scheduler updates will run tasks in the order created in the jobState object.
- also remove unused task arrays from job_state.go

### CODE REVIEW

File: scheduler/server/job_state.go [hunk 1]
```diff
@@ -84,9 +78,6 @@ func newJobState(job *domain.Job, jobClass string, saga *saga.Saga, taskDuration
 		JobKilled:                      false,
 		TimeCreated:                    time.Now(),
 		TimeMarker:                     time.Now(),
-		Completed:                      make(taskStateByTaskID),
```
Comment (JeanetteBruno) [inline scheduler/server/job_state.go:89]: these arrays are not used/needed.

File: scheduler/server/load_based_sched_alg.go [hunk 2]
```diff
@@ -111,10 +112,10 @@ func NewLoadBasedAlg(config *LoadBasedAlgConfig, tasksByJobClassAndStartTimeSec
...
 }
 
-// jobWaitingTaskIds map waiting task ids to the job state objects
-type jobWaitingTaskIds struct {
-	jobState       *jobState
-	waitingTaskIDs []string
+// jobWaitingTasks waiting task ids (in the order they should be started) for a job
+type jobWaitingTasks struct {
+	jobState     *jobState
+	waitingTasks []*taskState
```
Comment (JeanetteBruno) [inline scheduler/server/load_based_sched_alg.go:118]: we used to have taskIDs in waitingTaskIDs, and use the taskID to get the *taskState from the jobState's notRunning[] map.  Instead we now just store the *taskState's for tasks not started in waitingTasks (in the order they should be started) (see changes below)

File: scheduler/server/load_based_sched_alg.go [hunk 3]
```diff
@@ -271,18 +272,20 @@ func (lbs *LoadBasedAlg) initJobClassesMap(jobsByRequestor map[string][]*jobStat
 		for _, job := range jobs {
 			_, ok := jc.jobsByNumRunningTasks[job.TasksRunning]
 			if !ok {
-				jc.jobsByNumRunningTasks[job.TasksRunning] = []jobWaitingTaskIds{}
+				jc.jobsByNumRunningTasks[job.TasksRunning] = []*jobWaitingTasks{}
 			}
-			waitingTaskIds := []string{}
-			for taskID := range job.NotStarted {
-				waitingTaskIds = append(waitingTaskIds, taskID)
+			waitingTasks := []*taskState{}
```
Comment (JeanetteBruno) [inline scheduler/server/load_based_sched_alg.go:279]: build an array of not started *taskStates, maintaining the taskState order in job.Tasks

File: scheduler/server/stateful_scheduler.go [hunk 4]
```diff
@@ -99,6 +99,9 @@ const MaxPriority = domain.P2
 // TaskThrottle -
 //	   requestors will try not to schedule jobs that make the scheduler exceed
 //     the TaskThrottle.  Note: Sickle may exceed it with retries.
+// SchedAlg -
+//     config the configuration for the scheduling algorithm
```
Comment (dgdonut) [inline scheduler/server/stateful_scheduler.go]: Is this a typo?

File: scheduler/server/stateful_scheduler.go [hunk 5]
```diff
@@ -896,7 +907,7 @@ func (s *statefulScheduler) scheduleTasks() {
 				// Update the average duration for this task so, for new jobs, we can schedule the likely long running tasks first.
 				if err == nil || err.(*taskError).st.State == runner.TIMEDOUT ||
 					(err.(*taskError).st.State == runner.COMPLETE && err.(*taskError).st.ExitCode == 0) {
-					s.taskDurations[taskID].update(time.Now().Sub(tRunner.startTime))
+					s.taskDurations[s.getDurationTaskID(taskID)].update(time.Now().Sub(tRunner.startTime))
```
Comment (dgdonut) [inline scheduler/server/stateful_scheduler.go]: So this will still grow forever, just hopefully less?

**Comment by dgdonut** [comment]
> taskID from twitter scoot is the full pants build command (./pants test lint )

This is surprising. Was this always the case?

**Review by dgdonut — APPROVED** [review]
My inline comments are minor so approving.

File: scheduler/server/load_based_sched_alg.go (hunk 2 above)
Comment (anthonyjpratti) [inline scheduler/server/load_based_sched_alg.go:118]: Why slice of pointers?

File: scheduler/server/load_based_sched_alg.go [hunk 6]
```diff
@@ -126,7 +127,7 @@ type jobClass struct {
 	// jobsByNumRunningTasks is a map that bins jobs by their number of running tasks.  Given that the algorithm has
 	// determined it will start n tasks from class A, the tasks selected for starting from class A will give prefence
 	// to jobs with the least number of running tasks.
-	jobsByNumRunningTasks map[int][]jobWaitingTaskIds
+	jobsByNumRunningTasks map[int][]*jobWaitingTasks
```
Comment (anthonyjpratti) [inline scheduler/server/load_based_sched_alg.go]: Why slice of pointers?

File: scheduler/server/load_based_sched_alg.go (hunk 3 above)
Comment (anthonyjpratti) [reply scheduler/server/load_based_sched_alg.go:279]: Any comments you left on this PR should be comments in code

**Comment by dgdonut** [comment]
FYI, will probably collide with #506

File: scheduler/server/stateful_scheduler.go (hunk 4 above)
Comment (JeanetteBruno) [reply scheduler/server/stateful_scheduler.go]: removed

File: scheduler/server/load_based_sched_alg.go (hunk 2 above)
Comment (JeanetteBruno) [reply scheduler/server/load_based_sched_alg.go:118]: removed pointer nature

File: scheduler/server/load_based_sched_alg.go (hunk 6 above)
Comment (JeanetteBruno) [reply scheduler/server/load_based_sched_alg.go]: removed pointer nature

**Pull request merged**